kubectl apply: Deprecate --prune-whitelist in favor of --prune-allowlist

brianpursley opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Changes in kubectl apply --prune to support k8s [Inclusive Naming Initiative](https://www.cncf.io/announcements/2021/10/13/inclusive-naming-initiative-announces-new-community-resources-for-a-more-inclusive-future/):
* Deprecated the --prune-whitelist flag.
* Deprecated the PruneWhitelist field on ApplyFlags struct.
* Removed PruneWhitelist field (not used anywhere) from ApplyOptions struct.
* Added --prune-allowlist flag.
* Added PruneAllowlist field on ApplyFlags struct.
* Added unit tests for prune with allowlist

This commit also fixes a bug where the command would fail if you specified the same GVK multiple times for --allow-whitelist. Now it only attempts to prune the unique set of allowed GVKs.

#### Which issue(s) this PR fixes:
N/A

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Added a `--prune-allowlist` flag that can be used with `kubectl apply --prune`. This flag replaces and functions the same as the `--prune-whitelist` flag, which has been deprecated.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

brianpursley commented:
/hold 
for review

brianpursley commented:
Questions: 
1. Is it OK to deprecate -AND- hide the flag at the same time, or will that confuse people?
2. Assuming kubectl 1.29 is when the old flag can be removed (3 releases per year) if this makes it into 1.26. Is this correct?

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -224,7 +230,10 @@ func (flags *ApplyFlags) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringArrayVar(&flags.PruneAllowlist, "prune-allowlist", flags.PruneAllowlist, "Overwrite the default allowlist with <group/version/kind> for --prune")
+	cmd.Flags().StringArrayVar(&flags.PruneWhitelist, "prune-whitelist", flags.PruneWhitelist, "Overwrite the default whitelist with <group/version/kind> for --prune") // TODO: Remove this in kubectl 1.29 or later
+	cmd.Flag("prune-whitelist").Deprecated = "Use --prune-allowlist instead."
+	cmd.Flag("prune-whitelist").Hidden = true
```
I see there is a `cmd.Flags().MarkDeprecated` that does both deprecated and hidden at the same time. We use that in three places, which suggests doing both at once is indeed standard (makes sense to me as well--the flag will continue to work for folks already using it, but folks consulting the help won't newly discover it). Any reason not to use that method here?

brianpursley replied:
Ah thanks for that.  I wasn't aware of `MarkDeprecated`.  I'll change the code to use it.

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -93,7 +97,9 @@ type ApplyOptions struct {
-	PruneWhitelist          []string
+
+	// DEPRECATED: Not used
+	PruneWhitelist []string // TODO: Remove this in kubectl 1.29 or later
```
Is this because of the flag/option refactor initiative? Why leave the field in the struct if the behaviour is in fact broken if the consumer doesn't make the change?

brianpursley replied:
I don't know why this field was not being used (this PR isn't what makes it unused, it was already unused).

I could just go ahead and remove it, but I was thinking the deprecation period would apply not only to command-line usage, but also to any exported members.

I'm OK with (and prefer) removing it altogether, if that is OK to do.  I guess all that could happen is someone tries to depend on the newer kubectl and their build would fail if they were setting this field for some reason. It's probably a better outcome anyway, because it wouldn't be doing what they think it does.

KnVerey replied:
I agree with removing it then. I don't see a purpose in retaining something that already isn't functional--it can only mislead people.

KnVerey commented on staging/src/k8s.io/kubectl/pkg/util/slice/slice_test.go:
```diff
@@ -29,3 +30,44 @@ func TestSortInts64(t *testing.T) {
 	}
 }
+
+func TestToSet(t *testing.T) {
```
Love the unit tests! Would it be possible to add a regression test for the bug itself as well?

brianpursley replied:
I will write a unit test for that.

By the way, the existing prune unit test ([`TestApplyPruneObjects`](https://github.com/kubernetes/kubernetes/blob/cddbb0c56397448ac0489f0473a26601c1feece8/staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go#L663)) actually doesn't test prune at all.  If you remove the line [`cmd.Flags().Set("prune", "true")`](https://github.com/kubernetes/kubernetes/blob/cddbb0c56397448ac0489f0473a26601c1feece8/staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go#L697), the test still passes! :open_mouth:

brianpursley replied:
I added some unit tests for prune.

The test case that prevents a regression is `prune with allowlist specifying the same resource type multiple times should not fail`

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -66,7 +67,10 @@ type ApplyFlags struct {
-	PruneWhitelist []string
+
+	// DEPRECATED: Use PruneAllowlist instead
+	PruneWhitelist []string // TODO: Remove this in kubectl 1.29 or later
```
> Assuming kubectl 1.29 is when the old flag can be removed (3 releases per year) if this makes it into 1.26. Is this correct?

This is a good question. Strictly speaking, pruning is in alpha, which requires zero releases (the whitelist flag isn't marked as such, but it requires the --prune flag, which is marked, to do anything). But we've agreed that it needs to be treated at least as beta because of how long it has been around. Per https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli
> CLI elements of user-facing components (e.g. kubectl) must function after their announced deprecation for no less than:
> Beta: 3 months or 1 release (whichever is longer)

I think that means if we announce in 1.26, it needs to work at least in 1.27, but can be removed in 1.28? I'm not against being more generous though.

brianpursley replied:
I didn't realize beta features had a different deprecation timeline. I'll update the comment to say 1.28.

brianpursley committed:
kubectl apply: Deprecate --prune-whitelist in favor of --prune-allowlist

Changes in kubectl apply --prune to support k8s Inclusive Naming Initiative:
* Deprecated the --prune-whitelist flag.
* Deprecated the PruneWhitelist field on ApplyFlags struct.
* Removed PruneWhitelist field (not used anywhere) from ApplyOptions struct.
* Added --prune-allowlist flag.
* Added PruneAllowlist field on ApplyFlags struct.
* Added unit tests for prune with allowlist

This commit also fixes a bug where the command would fail if you specified
the sameGVK multiple times for --allow-whitelist. Now it only attempts to
prune the unique set of allowed GVKs.

KnVerey commented:
Thanks for adding all those tests! 

> By the way, the existing prune unit test (TestApplyPruneObjects) actually doesn't test prune at all. If you remove the line cmd.Flags().Set("prune", "true"), the test still passes! 😮

Looking at that test, I think it is completely redundant with the test above it. Digging through the history, it looks like it was introduced as a regression test in https://github.com/kubernetes/kubernetes/pull/85613, but it definitely does not cover that case either. It seems to me it would need at least two changes to cover that PR: it would need to apply non-namespaced objects exclusively and set expectations on `tf.FakeDynamicClient` instead of `tf.UnstructuredClient` asserting that the namespace specified in the flag is crawled. Out of scope for this PR though.

/lgtm

KnVerey commented:
@brianpursley I see this isn't auto-merging because you put a hold on it for review. Any further opinions you're looking for, or shall we remove that?

brianpursley commented:
> @brianpursley I see this isn't auto-merging because you put a hold on it for review. Any further opinions you're looking for, or shall we remove that?

@KnVerey I put the hold so it would wait for a review and not auto-merge

/unhold

brianpursley commented:
Rebased

pacoxu commented:
/lgtm

pacoxu commented:
/test pull-kubernetes-e2e-kind-ipv6
